README.txt
tournament
    by Douglas A. Seifert (doug+rubyforge at dseifert dot net)
    http://www.dseifert.net/code/tournament
    http://tournament.rubyforge.org/

== DESCRIPTION:

Small library, command line program and Rails web GUI for managing a NCAA
basketball tournament pool.

== FEATURES/PROBLEMS:

* Fully functional, self-service web GUI for collecting entries and
  generating reports
* Or use command line to add NCAA tournament pool entries and save them as YAML
* Run a possibilities report for determining who is likely to win
* Run other reports such as a leader board and entry report
* [DEPRECATED] Buggy, but functional, Shoes GUI included for creating
  the entries and updating the tournament results bracket.  Useful as an
  adjunct to the command line script.
* FIXME: Complete the test suite for the library and command line tool

== COMMAND LINE SYNOPSIS:

The tournament command line program is installed as 'pool'.  The library
has the 2009 NCAA tournament pre-configured.  If you were to use
this library for the 2010 NCAA tournament, code changes would be
necessary. FIXME: (add ability to read teams from a simple configuration
file).  For usage, just execute

   pool --help

For command specific usage, execute

   pool [command] --help

where [command] is one of the available commands described below. The
pool command saves state in a file called pool.yml by default.  This
can be overridden in all cases by using the --save-file option.

The pool manager would use this program as follows:

1. Choose a scoring strategy.  There are various scoring strategies
   that could be used.  The library comes pre-configured with 
   three scoring strategies:
   1. Basic scoring strategy: each correct pick is worth 2 X the round.
   2. Upset favoring strategy: each correct pick is worth a
      base amount per round plus the seed number of the winner.  As
      pre-configured, the base amounts per round are 3, 5, 11, 19, 30
      and 40 points.
   3. The Josh Patashnik strategy: each correct pick is worth the
      winner's seed number X a round multiplier.  The multipliers
      are 1, 2, 4, 8, 16 and 32 points.
   4. Tweaked Josh Patashnik strategy: each correct pick is worth the
      winner's seed number X a round multiplier.  The multipliers
      are 1, 2, 4, 8, 12 and 22 points.
   4. Constant Value strategy: each correct pick is worth exactly
      one (1) point, regardless of round.
   If your scoring strategy is not one of the above, you will have to
   add a class to the ScoringStrategy module, in file
   lib/tournament/scoring_strategy.rb.

2. Create a directory to hold the pool data and change to it

3. Initialize the pool

       pool setup [--scoring=upset]

   Use the --scoring argument to change from the default basic scoring
   strategy.  If the basic strategy is ok, the --scoring argument is
   not required.

   As mentioned above, unless overridden by using the --save-file
   option, the pool will save itself to the file 'pool.yml'

4. Set the entry fee and payout amounts
   
       pool fee 10
       pool payout 1 80
       pool payout 2 20
       pool payout last 10 -C

   The above commands say that each entry fee is 10 units (this is all
   for fun, not profit, right?) and that the 1st place finisher would
   receive 80% of the total payout, the 2nd place finisher would 
   receive 20% of the total payout and the last place finisher would
   receive 10 units back (would get her entry fee back).  No error
   checking is done with this.  FIXME: Add error checking.

5. Export a tournament entry YAML file

       pool dump

   This will save the tournament entry file as tournament.yml unless
   the --entry option is used to override it.

6. Create entries.  You can use the included buggy GUI (see below),
   or edit YAML files by hand.

7. Import the entry YAML files into the pool

       pool entry --add=path/to/entry.yml

8. As games progress, update the tournament.yml file, again using the GUI or
   editing the YAML file by hand.  Then update the pool with the new
   pool YAML file

       pool update

9. Run reports
  
       pool report [final_four|entry|region|leader|score]

   The final four report can only be run once the final four teams have
   been determined.

10. After about 22 teams are left, run a possibility report.  This report will
    run through all the remaining ways the tournament can come out and
    calculate the chance to win for each player.  The chance to win
    is defined as the percentage of possibilities that lead to that player
    coming out on top in the pool.  With more than about 22 teams left
    (YMMV), this report could take months to run. FIXME (Investigate
    possibly using EC2 or something to spread the load around, or 
    otherwise optimize the possibility checking algorithm)

        pool report possibility

== WEB GUI:

A Rails web application is available if you don't want to use the
command line to manage your pool.

=== INSTALLING THE WEB GUI:

The web application can be installed by running the pool command
as follows

   pool install_webgui --web-dir=/path/to/directory [options]

The above command will copy the Rails app to the specified directory.

There are several options you can provide in addition to --web-dir
to control how the application is installed:

* Human readable site name.  This appears as the title tag content
  of pages in the site and is also used in the subject line of any
  emails sent by the site (as during user registration).

    --site-name="Site Name"  (Default: 'Tournament')

* Relative url root.  If you will be installing the pool site as
  a relative url on another virtual host, use this switch.  You
  will have to configure you virtual host to route requests to this
  path to Rails.  This is ridiculously easy if you are using
  modrails and Apache.  Example:

    --relative-root=/my_pool  (Default: empty, no root is set)
 
* Administrator email address.  The web GUI will send emails
  when users register for creating entries in a pool. The following
  sets the from email address on the emails that are sent.

    --admin-email=admin

* Email server information.  Either edit the config/initializers/pool.rb
  file after installation, or provide the following options to configure
  a SMTP server available on your domain.

    --email-server=smtp.myisp.com
    --email-port=25
    --email-domain=mydomain.com
    --email-user=myuser
    --email-password=mypass
    --email-auth=login|plain|cram_md5

  See http://guides.rubyonrails.org/action_mailer_basics.html#_action_mailer_configuration
  for more info on how to configure a Rails app for sending email.

* The web GUI has the ability to print bracket entries by generating
  a pdf styled using the web site's bracket stylesheet.  It uses a third
  party tool called Prince XML to do this.  You are not allowed to
  use this on a server without paying a license fee, although you
  can download a trial version for personal use.  It's your call
  whether or not you want to use this.  Please see http://www.princexml.com/
  for more details.
 
    --use-princexml=/full/path/to/prince

  If prince is not available on the path you specify, the princexml
  distribution will be downloaded and installed using the distribution's
  install.sh script.  In order to do this, the tar program must be
  available on your installation system.

* If you use the --use-princexml option, the install script needs to write
  files to a temp directory, /tmp by default. Use the --tmp-dir option to
  change this default.  If the specified temp dir does not exist, it will
  be created.

    --tmp-dir=/path/to/tmp/dir

=== POST-INSTALLATION:

Before being able to run the web gui for the first time, you have to
generate the web site authorization keys, prepare the sqlite database
and create an admin account.  Change to the website installation
directory and perform the following steps.

1. Generate the web site authorization keys

     RAILS_ENV=production rake auth:gen:site_key

2. Prepare the sqlite database

     RAILS_ENV=production rake db:migrate

3. Create the admin account

     RAILS_ENV=production rake "admin:create[login,email,name,password]"

   In the above command, substitute "login" for the desired admin user's login
   name, "email" for the administrators email address, "name" for the 
   admin user's name (eg, "Joe Admin"), and "password" for the desired
   admin account password

=== UPDATING THE WEB GUI:

If the tournament gem is updated, you can pull in the changes as follows:

1. Update your tournament gem

     sudo gem update tournament

2. Rerun the same pool install_webgui command you used to install originally

3. Pull in any released db migrations:

     cd $install_dir; RAILS_ENV=production rake db:migrate

4. Reload your web server configuration (nginx, apache, etc.)

=== USING THE WEB GUI:

Load the entry page in your brower.  Log in as the admin user you 
configured during installation.  Click on on the 'All Pools' link on
the right sidebar.  Create a new pool and fill in the information.
Each time you save this form, a new blank payouts line will be added
so that you can configure as many payouts as you desire.

After the basic pool configuration is set up, click on the 'Teams'
link in the right sidebar.  You are presented with four region
brackets to fill in.  Keep in mind that the pink region champs will
play each other in the final four and the light blue region champs
will play each other in the final four so you can get the bracket 
right.  The web application is preconfigured with over 300 NCAA
schools.  The team name fields are auto-complete fields -- type in
a few letters and pause and you will be presented with a list of
matching teams.  The Short Name field should be a three letter
abbreviation for the team.  The abbreviations have to be unique
across the entire tournament field.

Once the teams are configured, go back to the pool basic information
form, click the Active check box and save the form.  The pool is now
ready for entries to be added to it.  Invite your friends to join the pool
by giving them the url for the entry page.  They will be asked to
register.  After registering and logging in, they will be able to
submit entries to your pool.

As the tournament progresses, use the 'Tournament Bracket' link
on the right sidebar to record the winning teams.  

Use the report links to run reports, etc.

==== POSSIBILITY REPORT

After about 21 teams are left in the tournament, you can run the
possibility report.  This report runs through every possible way the
tournament can come out and ranks each entry against the possiblity.
The report lists the "chance to win" for each entry.  The chance
to win is the percentage of possible outcomes that would result in
that entry coming in first.

The possibility report requires that a rake task be run on the
web server.  It is very processor intensive and can take a long
time to complete.  To generate the possibility report data file,
run the following command from the web gui install directory
on the server:

   RAILS_ENV=production rake report:possibilities

== SHOES GUI:

A GUI for filling out tournment bracket entries is included and is run
by executing the program "picker".  If no argument is included, a blank
entry is started.  It can be filled in and saved using one of the buttons
at the top of the screen.  The entry is saved as a YAML file for a
Tournament::Entry object.   The picker program optionally takes one
argument, the path to a Tournament::Entry YAML file.  It will open
with the provided entry's picks pre filled in.

The GUI also may be used for keeping the NCAA tournament entry YAML file
up to date:

   picker tournament.yml

The GUI works as long as you don't try to go back and change games
from played to unknown outcome.
 
== REQUIREMENTS:

* main (2.8.0)

== WEB GUI REQUIREMENTS:

* rails (2.2.2)
* rake (0.8.3)
* sqlite3-ruby (1.2.4)

== SHOES GUI REQUIREMENTS:

Verified working on
* shoes raisins v1134

== INSTALL:

* sudo gem install tournament
* Download the tgz file and run it locally.  If the latter
  option is taken, the tournament/bin directory must be in
  the path.

== LICENSE:

(The MIT License)

Copyright (c) 2008

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 5.0.0 / 2012-03-08
* 2012 bracket.

== 4.2.0 / 2011-03-20
* Add ability to run statistics for the possibility report in more than one
  process.
* Add link to possibility report on pool index page.
* Pretty formatting for large numbers.

== 4.1.0 / 2011-03-17
* Bug fix change team name and some cosmetic changes

== 4.0.2 / 2011-03-14
* Add ability to change the name of a team after tourny has started.  

== 4.0.0 / 2011-03-14
* 2011 bracket.
* Upgrade to Rails 2.3.11.

== 3.3.3 / 2010-03-21
* Show link to possibility report if stats file exists.

== 3.3.2 / 2010-03-17
* More regressions: Fix Tourny Bracket link when logged in as pool owner.
  Fix create entry to set the user_id so the new permissions checking actually
  works.

== 3.3.1 / 2010-03-16
* Emergency bug fix.

== 3.3.0 / 2010-03-16
* Final 2010 bracket.
* Fix permissions issue making tournament bracket seem editable to all.
* Show login name in leader board.

== 3.2.2 / 2010-03-14
* Fix initialization of tournament entry.

== 3.2.1 / 2010-03-14
* Fix teams selection controller to not care about the order of
  input seedings.

== 3.2.0 / 2010-03-14
* Fix team selection page to not jump around when using arrow keys
  with the autocomplete pop ups.
* Allow switch between seed and matchup order in team selection page.
* Minor cosmetic formatting changes.

== 3.1.1 / 2010-03-13
* Bugfixes for showing add entry links.  Makes sure participants can
  add entries once the teams are set.
* Verbiage for pools that are over.

== 3.1.0 / 2010-03-12
* Account for multiple pools when saving/loading possibility yaml file.
* Update restful_authentication to latest version.
* Add ability to reset passwords.

== 3.0.3 / 2010-03-11
* Fix bug with bad string encoding and prince xml for pdf generation.

== 3.0.2 / 2010-03-10
* Fix bug caused by change in semantics of Array#to_s in ruby 1.9.1

== 3.0.1 / 2010-03-10
* Fix bug with relative_url_root in pool initializer.
* git housekeeping, add .gitignore
* Configure bones to ignore more files during gem package

== 3.0.0 / 2010-03-06
* NCAA 2010 Tournament initial release.

== 2.6.0 / 2009-03-29
* Pretty HTML Final Four report.

== 2.5.2 / 2009-03-29
* Really fix manifest problems.
* Adjust webgui install instructions to account for
  config/initializers/site_key.rb creation (don't reuse the
  dev site_keys).

== 2.5.1 / 2009-03-26
* Attempt to fix manifest problems

== 2.5.0 / 2009-03-23
* Add rake task for generating possibility report data file.
* Bugfix for final four report: work around fact that web gui
  sets things as strings that it should not.

== 2.4.0 / 2009-03-23
* Hook up possibility report.
* Limit recap mails to pool participants.

== 2.3.0 / 2009-03-20
* Add ability for admin to send emails to participants.  Needs work.

== 2.2.1 / 2009-03-19
* Bug fixes and minor display enhancement.

== 2.2.0 / 2009-03-18
* Final update for 2009 NCAA tournament
* Fix a couple of anchor tags.

== 2.1.2 / 2009-03-16
* Add two new scoring strategies.
* Another release forthcoming when play-in game is decided.

== 2.1.1 / 2009-03-16
* Fix bug when changing entry name after it has been completed
  and therefore added to the backing Tournament::Pool object
* Another release forthcoming when play-in game is decided.

== 2.1.0 / 2009-03-15
* Update for 2009 NCAA tournament seedings
* Update for 2009 NCAA tournament seedings
* Leader board report is HTML instead of text
* Minor formatting changes for main page
* Another release forthcoming when play-in game is decided.

== 2.0.0 / 2009-03-12
* Release Rails GUI for self-service entries and admin
* Make reports account for tournament tie breaker.
* YAML format change from version 1.X.  Old pool data won't work
  with this version.
* Added Josh Pasternak scoring strategy.  Correct picks are worth
  the winning team seed number * a per round multiplier.  See
  http://blogs.tnr.com/tnr/blogs/the_plank/archive/2008/03/27/ncaa-pool-scoring-system-rant.aspx
  for details.
* Many library improvements and refactorings.
* Another release forthcoming with final 2009 NCAA tournament seedings forthcoming

== 1.1.1 / 2008-07-07
* Update rake tasks to bones-2.0.2

== 1.1.0 / 2008-03-29

* Add ability to configure entry fee and payout amounts
* Change final four report to use the entry fee and payout amount configuration
* Fix bug with cached @number_of_outcomes in Tournament::Pool class.
* Add to the possibility report a listing of each entry that could place first
  that shows frequency of which team would have to be champion for 
  the entry to place first.

== 1.0.0 / 2008-03-26

* Change the pool command line script usage to make it simpler (I hope)
* Add a new report: final_four
* FIXME: Final four report assumes payouts go to 1st and 2nd and last places.
  This should be configurable.

== 0.0.3 / 2008-03-23

* Bugfix the picker wrapper script.  Getting used to bones/rubyforge.

== 0.0.2 / 2008-03-23

* Changed the picker wrapper script to ruby because it seems that
  that is what rubyforge tools expect.

== 0.0.1 / 2008-03-23

* Initial Release



webgui/db/schema.rb
# This file is auto-generated from the current state of the database. Instead of editing this file, 
# please use the migrations feature of Active Record to incrementally modify your database, and
# then regenerate this schema definition.
#
# Note that this schema.rb definition is the authoritative source for your database schema. If you need
# to create the application database on another system, you should be using db:schema:load, not running
# all the migrations from scratch. The latter is a flawed and unsustainable approach (the more migrations
# you'll amass, the slower it'll run and the greater likelihood for issues).
#
# It's strongly recommended to check this file into your version control system.

ActiveRecord::Schema.define(:version => 20100312053540) do

  create_table "entries", :force => true do |t|
    t.string   "name",       :limit => 64,                    :null => false
    t.binary   "data"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "tie_break"
    t.integer  "user_id",                  :default => 1,     :null => false
    t.integer  "pool_id"
    t.boolean  "completed",                :default => false, :null => false
  end

  add_index "entries", ["pool_id"], :name => "index_entries_on_pool_id"
  add_index "entries", ["user_id"], :name => "index_entries_on_user_id"

  create_table "pools", :force => true do |t|
    t.string   "name",                          :null => false
    t.binary   "data"
    t.boolean  "started",    :default => false, :null => false
    t.datetime "starts_at"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "user_id"
    t.boolean  "active",     :default => false, :null => false
  end

  create_table "regions", :force => true do |t|
    t.integer  "pool_id"
    t.string   "name"
    t.integer  "position"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "roles", :force => true do |t|
    t.string   "name",       :limit => 32
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "position",                 :default => 0, :null => false
  end

  create_table "roles_users", :id => false, :force => true do |t|
    t.integer "role_id"
    t.integer "user_id"
  end

  create_table "seedings", :force => true do |t|
    t.integer  "pool_id"
    t.integer  "team_id"
    t.string   "region"
    t.integer  "seed"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "teams", :force => true do |t|
    t.string   "name"
    t.string   "short_name"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "users", :force => true do |t|
    t.string   "login",                     :limit => 40
    t.string   "name",                      :limit => 100, :default => ""
    t.string   "email",                     :limit => 100
    t.string   "crypted_password",          :limit => 40
    t.string   "salt",                      :limit => 40
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "remember_token",            :limit => 40
    t.datetime "remember_token_expires_at"
    t.string   "activation_code",           :limit => 40
    t.datetime "activated_at"
    t.string   "password_reset_code",       :limit => 40
  end

  add_index "users", ["login"], :name => "index_users_on_login", :unique => true

end
